### ISSUE TITLE
`produce()` ignores `power` for buildings

### ISSUE THREAD

**Issue by hexagonrecursion** [colobot/colobot#1748]
When I tried to spawn a DefenseTower with a power cell via `produce`, the tower spawned without a power cell

```c++
extern void object::New()
{
	produce(position, 0, DefenseTower, "", 1);
}
```

game version: gold-0.2.1

Thanks to Zemo/Jendrus on discord for discovering this

**Comment by melex750** [comment]
Yes, it's an easy fix too if you follow how the bots get their power cell.

**Comment by hexagonrecursion** [comment]
1. Robots create power cells here:
  https://github.com/colobot/colobot/blob/8a40853234a054a73753a3afe7e80ff2e209f43f/colobot-base/src/object/motion/motionvehicle.cpp#L1060-L1082
1. PowerPlant and PowerStation already respect the `power` parameter - only ResearchCenter and DefenseTower need to be fixed
  https://github.com/colobot/colobot/blob/8a40853234a054a73753a3afe7e80ff2e209f43f/colobot-base/src/object/subclass/base_building.cpp#L537
  https://github.com/colobot/colobot/blob/8a40853234a054a73753a3afe7e80ff2e209f43f/colobot-base/src/object/subclass/base_building.cpp#L354

**Comment by melex750** [comment]
> PowerPlant and PowerStation already respect the power parameter...

Does this mean you can choose the energy level a PowerStation starts with? (In it's own built-in PowerCell)?

Also want to mention NuclearPlant as relevant here..

**Comment by hexagonrecursion** [comment]
> > 
> 
> Does this mean you can choose the energy level a PowerStation starts with? (In it's own built-in PowerCell)?

    produce(position, 0, PowerStation, "", 0.5);

Result:

![image](https://github.com/user-attachments/assets/27fe05ab-620b-4fd4-af5a-f7685eb481cd)

**Comment by hexagonrecursion** [comment]
> Also want to mention NuclearPlant as relevant here..

Why? We aren't discussing a general mechanism to pre-place arbitrary objects into arbitrary slots. Mostly just placing pre-powered defence towers. I think NuclearPlant is *not* relevant because it is not *powered by* the NuclearCell.

**Comment by melex750** [comment]
> ...a general mechanism to pre-place arbitrary objects into arbitrary slots.

Yeah, it's not directly relevant. It's also not an arbitrary slot on NuclearPlant, it is the ```energyCell``` slot.

> general mechanism

We could use numbers greater than 100 to place other objects in that slot. Maybe a number that represents color like underground resources bitmap... or object_type number.

**Comment by hexagonrecursion** [comment]
> We could use numbers greater than 100 to place other objects in that slot. Maybe a number that represents color like underground resources bitmap... or object_type number.

The negative numbers are unused (except for -1). We can use `produce(..., -TitaniumOre)` to make something with TitaniumOre in the power slot.

**Comment by hexagonrecursion** [comment]
> I think in the long run we might want to change how inserting objects works. Current `scene` format supports placing batteries in objects, as there weren't many other objects you'd want to insert inside. But with [slots implemented](https://github.com/colobot/colobot/pull/1444) and possibly more objects possible to add into game through mods, I'm thinking that new format should be a target, rather than complicating current format.
> 
> There was an idea to move `scene` files to JSON (or alike, like XML or YAML) format, with current `scene` kept for backwards compatibility. Since JSON allows nesting (unlike current format) we could just define the object inside parent object and attach it to any given slot with nesting, instead of dealing with magic numbers. This way we could easily create Grabbers with given battery in `energyCell` slot and any (possibly even inserted by mod) object in the arm.
> 
> This conversion probably still needs some discussion about implementation technicalities under separate issue/discussion. I'm just throwing the idea here to think if expanding current format is work-effective, or we should limit the scope of this issue to fixing a bug and wait with implementing new features in more flexible format.

We do have CreateFret, CreatePower and CreateSlotObject. Kinda clunky, but it works. At a glance it appears that they are currently only supported in save files, but we could change that.

https://github.com/colobot/colobot/blob/8a40853234a054a73753a3afe7e80ff2e209f43f/colobot-base/src/level/robotmain.cpp#L4749-L4769

**Comment by tomangelo2** [comment]
Just for clarify (so it won't haunt my mind) - while the issue is about CBot function `produce()` and my comment is about level format, I think that we should keep passing parameters like these similar across engine.
Current `produce()` works somehow similar to current `Scene`, as in you have baked in what parameters you can change in object you can produce. If we'd move with level files to new format, maybe we could also make produce (or it's more robust alternative, like `search()` vs `radar()`) more flexible and similar?
Example: `produce(prototype)`, where prototype is an array/object, where every property is possible to edit. Kind of like existing `object` type, but every field is writable instead of read-only. So if you want to create a robot that's already a bit damaged, discharged and holding specific object, you define an object variable and set every detail you need.
```cpp
prototype bot = new Prototype();
bot.category=WingedGrabber;
bot.shieldLevel = 0.6;
bot.slot0.category = NuclearCell;
bot.energyCell.energyLevel = 0.2; //energyCell being an alias for "slot0"
bot.slot1.category = BlackBox;
bot.load.shieldLevel = 0.9; //like above, load is an alias for "slot1"
//number of slots defined in object definition, either in engine or in separate files
//other properties are left as default
//alternatively, create a "prototype" for BlackBox with it's own properties, and use it with:
//bot.load = bb; //bb is a "prototype" variable, just like bot
//this way you can define it once and use it anywhere
[...]
produce(bot); //or create(bot); if we're going to create separate function instead of overloading existing one
```

That's why I've mentioned levels, so creating same thing in scene (and in save files, since they share a syntax) would have similar syntax (I'm using JSON, but the idea should work in similar formats as well):
```json
{
  "Objects":[
    {"category":"WingedGrabber", "shieldLevel":0.6, "slots":[
      {"category":"NuclearCell", "energyLevel":0.2},
      {"category":"BlackBox", "shieldLevel":0.9}
      ]
    }
  ]
}
```

So to rephrase what I messily wrote earlier: instead of overloading existing simple mechanisms to accommodate new features, maybe it might be better to move advanced features to new systems and keep existing ones simple and backwards compatible.

**Comment by hexagonrecursion** [comment]
I have an idea. I was already considering adding the ability to parse and create JSON to CBOT for other reasons. We could just add an overload of `produce()` that takes exactly the same JSON as level files do (assuming we decide to use JSON or YAML in level files)
